worker/src/RTC/ActiveSpeakerObserver.cpp

```diff
@@ -0,0 +1,560 @@
+		{
+			Update();
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerResumed(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+
+			rtpObserver.speaker->paused = false;
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerPaused(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+
+			rtpObserver.speaker->paused = true;
+		}
+	}
+
+	void ActiveSpeakerObserver::ReceiveRtpPacket(RTC::Producer* producer, RTC::RtpPacket* packet)
+	{
+		MS_TRACE();
+
+		if (IsPaused())
+			return;
+
+		uint8_t volume;
+		bool voice;
+
+		if (!packet->ReadSsrcAudioLevel(volume, voice))
+			return;
+
+		auto it = this->mapProducerSpeaker.find(producer->id);
+
+		if (it != this->mapProducerSpeaker.end())
+		{
+			auto& rtpObserver = it->second;
+			uint64_t now      = DepLibUV::GetTimeMs();
+
+			rtpObserver.speaker->LevelChanged(volume, now);
+		}
+	}
+
+	void ActiveSpeakerObserver::Paused()
+	{
+		MS_TRACE();
+
+		this->periodicTimer->Stop();
+	}
+
+	void ActiveSpeakerObserver::Resumed()
+	{
+		MS_TRACE();
+
+		this->periodicTimer->Restart();
+	}
+
+	void ActiveSpeakerObserver::OnTimer(Timer* timer)
+	{
+		MS_TRACE();
+
+		Update();
+	}
+
+	void ActiveSpeakerObserver::Update()
+	{
+		MS_TRACE();
+
+		uint64_t now              = DepLibUV::GetTimeMs();
+		uint64_t levelIdleTimeout = LevelIdleTimeout - (now - lastLevelIdleTime);
+
+		if (levelIdleTimeout <= 0)
+		{
+			if (lastLevelIdleTime != 0)
+			{
+				TimeoutIdleLevels(now);
+			}
+			lastLevelIdleTime = now;
+		}
+
+		if (mapProducerSpeaker.size() > 0 && CalculateActiveSpeaker())
+		{
+			json data          = json::object();
+			data["producerId"] = this->dominantId;
+
+			Channel::ChannelNotifier::Emit(this->id, "dominantspeaker", data);
+		}
+	}
+
+	bool ActiveSpeakerObserver::CalculateActiveSpeaker()
+	{
+		MS_TRACE();
+
+		std::string newDominantId;
+		int32_t speakerCount = mapProducerSpeaker.size();
+
+		if (speakerCount == 0)
+		{
+			newDominantId = "";
+		}
+		else if (speakerCount == 1)
+		{
+			auto it = mapProducerSpeaker.begin();
+
+			newDominantId = it->second.producer->id;
+		}
+		else
+		{
+			Speaker* dominantSpeaker =
+			  (dominantId.empty()) ? nullptr : mapProducerSpeaker[dominantId].speaker;
+
+			if (dominantSpeaker == nullptr)
+			{
+				auto item       = mapProducerSpeaker.begin();
+				newDominantId   = item->first;
+				dominantSpeaker = item->second.speaker;
+			}
+			else
+			{
+				newDominantId = "";
+			}
+
+			dominantSpeaker->EvalActivityScores();
+			double newDominantC2 = C2;
+
+			for (auto it = mapProducerSpeaker.begin(); it != mapProducerSpeaker.end(); ++it)
+			{
+				Speaker* speaker     = it->second.speaker;
+				const std::string id = it->second.producer->id;
```

Comment:
Let's avoid string copy:

```suggestion
				Speaker* speaker      = it->second.speaker;
				const std::string& id  = it->second.producer->id;
```